# Algoritma Genetika Untuk Optimasi Fungsi

## Permasalahan
Mencari nilai $x_1, x_2, x_3$ yang dapat meminimalkan fungsi
$$f(x_1,x_2,x_3)=(x_1+2x_2-7)^2+(2x_1+x_2-5)^2+(x_2+x_3)^2$$
dengan $-10<x_1,x_2,x_3<10$ menggunakan algoritma genetika, dimana probabilitas terjadinya crossover 0.99, probabilitas terjadinya mutasi 0.98 dan banyaknya kromosom adalah 100

## Pseudocode Algoritma Genetika
**START** <br>
Membentuk kromosom <br>
mendefinisikan fungsi fitnes <br>
Membentuk populasi awal ($n$ kromosom) <br>
Menentukan $p_xc$, $p_mm$, dan $N$ <br>
**for** $i=1$ **to** $N$ <br>
$\hspace{1cm}$ **for** $j=1$ **to** $n$ <br>
$\hspace{2cm}$ **Seleksi** <br>
$\hspace{2cm}$ **Crossover** <br>
$\hspace{2cm}$ **Mutasi** <br>
$\hspace{1cm}$ **end for** <br>
$\hspace{1cm}$ Gabung semua kromosom <br>
$\hspace{1cm}$ Pilih $n$ kromosom terbaik (populasi baru) <br>
**end for** <br>
**STOP**

## Implementasi

#### Import library

In [1]:
from geneticalgorithm import GeneticAlgorithm

#### Masukan-masukan

In [2]:
# Jumlah kromosom
n = 100
# Jumlah variabel
m = 3

# Batas bawah dan atas
min_value = -10
max_value = 10

# Propabilitas corssover dan mutasi serta jumlah generasi
pc = 0.99
pm = 0.98
N = 100

#### Hitung hasil

In [3]:
gen_al = GeneticAlgorithm(n_chrom=n, chrom_length=m,
      min_value=min_value, max_value=max_value,
      p_cross=pc, p_mut=pm, max_generate=N)
gen_al.solve()

x1         1.078962
x2         2.890803
x3        -2.714742
fitness    0.052813
Name: 0, dtype: float64

## Kesimpulan
Diperoleh nilai $x_1=1.08,x_2=2.89,x_3=-2.71$. Sedangkan nilai eksaknya adalah $x_1=1,x_2=w,x_3=-3$. Jadi algoritma genetika cukup bagus untuk menentukan hasilnya, tergantung parameter-parameter yang digunakan bisa mempengaruhi hasil yang didapatkan